src/main/redux/sagas/app.ts

```diff
@@ -50,4 +61,101 @@ export function* init() {
         },
     );
 
+    app.on("will-quit", () => {
+
+        debug("#####");
+        debug("will-quit");
+        debug("#####");
+    });
+
+}
+
+export function exit() {
+    return spawn(function*() {
+
+        const beforeQuitEventChannel = getBeforeQuitChannel();
+        const windowAllClosedEventChannel = getWindowAllClosedEventChannel();
+        const quitEventChannel = getQuitEventChannel();
+        let shouldExit: boolean = process.platform !== "darwin" || IS_DEV;
+
+        /*
+        // events order :
+        - before-quit
+        - window-all-closed
+        - quit
+        */
+
+        yield takeSpawnEveryChannel(
+            beforeQuitEventChannel,
+            (e: Electron.Event) => {
+
+                // track ctrl-q/command-q
+                shouldExit = true;
+
+                e.preventDefault();
+
+                const libraryWin = getLibraryWindowFromDi();
+                libraryWin.close();
+
+                debug("#####");
+                debug("#####");
+                debug("#####");
+
+                debug("before-quit");
+
+                debug("#####");
+                debug("#####");
+                debug("#####");
+            },
+        );
+
+        yield takeSpawnEveryChannel(
+            windowAllClosedEventChannel,
+            function*() {
+
+                debug("#####");
+                debug("#####");
+                debug("#####");
+
+                debug("window-all-closed");
+
+                debug("#####");
+                debug("#####");
+                debug("#####");
+
+                // clear session in r2-navigator
+                yield all([
+                    call(clearSessions),
+                    call(needToPersistState),
+                ]);
+
+                yield put(streamerActions.stopRequest.build());
+
+                yield race({
+                    a: take(streamerActions.stopSuccess.ID),
+                    b: take(streamerActions.stopError.ID),
+                });
```

Comment:
In the Electron app lifecycle, `window-all-closed` would typically be used to differentiate a Mac app which can "survive" without any active window. As good practice, this event handler should not contain teardown logic, it should instead simply `app.quit()` (or not, for example on Mac). Then, the `[will|before]quit` events can be used to capture the teardown phase, for all platforms (Linux, Windows, Mac).
I would therefore recommend migrating any teardown logic (i.e. graceful shutdown of internal services) to one of the `quit` events. This way, all scenarios are covered: user closing windows one by one until there are no more, or user closing library view while there are reader windows open (which results in cascade-closing all reader windows), or force-closing using Apple-Q / Alt-F4.